domain/network/state/space.go

```diff
@@ -0,0 +1,247 @@
+    ON space.uuid = subnet.space_uuid
+    LEFT JOIN provider_subnet
+    ON subnet.uuid = provider_subnet.subnet_uuid
+    LEFT JOIN provider_network_subnet
+    ON subnet.uuid = provider_network_subnet.subnet_uuid
+    LEFT JOIN provider_network
+    ON provider_network_subnet.provider_network_uuid = provider_network.uuid
+    LEFT JOIN availability_zone_subnet
+    ON availability_zone_subnet.subnet_uuid = subnet.uuid
+    LEFT JOIN availability_zone
+    ON availability_zone_subnet.availability_zone_uuid = availability_zone.uuid`
+
+// GetSpace returns the space by UUID.
+func (st *State) GetSpace(
+	ctx context.Context,
+	uuid string,
+) (*network.SpaceInfo, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Append the space uuid condition to the query only if it's passed to the function.
+	q := retrieveSpacesStmt + " WHERE space.uuid = $M.id;"
+
+	s, err := sqlair.Prepare(q, Space{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var rows Spaces
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{"id": uuid}).GetAll(&rows))
+	}); err != nil {
+		return nil, errors.Annotate(err, "querying spaces")
+	}
+
+	if len(rows) == 0 {
+		return nil, errors.NotFoundf("space %q", uuid)
+	}
+
+	return &rows.ToSpaceInfos()[0], nil
+}
+
+// GetAllSpaces returns all spaces for the model.
+func (st *State) GetAllSpaces(
+	ctx context.Context,
+) (network.SpaceInfos, error) {
+
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	s, err := sqlair.Prepare(retrieveSpacesStmt, Space{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", retrieveSpacesStmt)
+	}
+
+	var rows Spaces
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, s).GetAll(&rows))
+	}); err != nil {
+		return nil, errors.Annotate(err, "querying all spaces")
+	}
+
+	return rows.ToSpaceInfos(), nil
+}
+
+// UpdateSpace updates the space identified by the passed uuid.
+func (st *State) UpdateSpace(
+	ctx context.Context,
+	uuid string,
+	name string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	q := `
+UPDATE space
+SET    name = ?
+WHERE  uuid = ?;`
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		if _, err := tx.ExecContext(ctx, q, name, uuid); err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+}
+
+// DeleteSpace deletes the space identified by the passed uuid.
+func (st *State) DeleteSpace(
+	ctx context.Context,
+	uuid string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteSpaceStmt := "DELETE FROM space WHERE uuid = ?;"
+	deleteProviderSpaceStmt := "DELETE FROM provider_space WHERE space_uuid = ?;"
+	updateSubnetSpaceUUIDStmt := "UPDATE subnet SET space_uuid = NULL WHERE space_uuid = ?;"
```

Comment:
so when we delete a space, then its previously associated subnet becomes part of `alpha`?

Reply:
The logic for checking whether a space can be removed (and whether subnets can change spaces) needs to be pushed into the SQL, but we can't do that yet because we need to check whether a space is used by configuration or bindings.

If it *was* valid to delete a space, the subnets in it would go into `alpha`, yes.

Reply:
Oof. I approved a patch with incorrect behaviour. See [here](https://github.com/juju/juju/pull/13753/files#diff-063734135ca5f0d1aa513ed10712426f1a0b4bce2914cdb6ccaefe45a92c7d3d) for the change _from_ updating to `alpha`.

Reply:
done, see a0f2466291